Update lxc-net to support nftables

pablofsf opened the pull request:
Closes #3093
Closes #3602

Add support for nftables firewall rules if `nft` command line
interface is available in the system

I have tested this in a local machine (without ipv6) with nftables and seems to be working perfectly fine. 

Note that I have removed the `CHECKSUM` rule in nft, as the automatic rule translation creates something completely meaningless: `oifname "lxcbr0" meta l4proto udp udp dport 68 counter packets 0 bytes 0 # CHECKSUM fill`, that in addition never sees any packages.

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

stgraber commented:
Actually going to let @tomponline review this one. He knows more about nft than I do :)

tomponline commented:
@stgraber should we just use nft if its present or should we require a config/env var to trigger the change of tooling?

stgraber commented:
Doing a complex check like in LXD is a bit impractical, but so is just assuming it will work if present.

So maybe a middle ground is:
 - Allow the user to force a backend through a config key (LXC_NET_NFT=0 to disable?)
 - Check if binary exists
 - Check if `nft list ruleset` succeeds

If it's not manually disabled, the binary is found and list succeeds, then we go with nft.

pablofsf commented:
@stgraber @tomponline Makes sense to add some extra checks. I've updated the checks, used `LXC_USE_NFT` and added the variable to `lxc.spec.in` so that users are aware of its use.

tomponline commented on config/init/common/lxc-net.in:
```diff
@@ -51,6 +55,52 @@ _ifup() {
+
+start_nftables() {
+    start_ipv6
+    # It is not possible to delete rules in nftables using directly the same
```
nftables doesn't have any built in tables & chains like iptables does. However because there is an iptables shim that populates nftables, it can sometimes look like nftables has the the `filter` table already.

So for example, on Ubuntu 20.04, running:

```
sudo nft list ruleset
```

Produces no output, and running:

```
sudo nft add chain ip6 nat lxc-nat
```

Fails with:

```
Error: No such file or directory
add chain inet filter lxc-filter
```

This is because the inet filter table doesn't exist.

My recommendation is that you test this script by resetting the nftables state first using `sudo nft flush ruleset`.

tomponline replied:
Also on that point, is there any reason you aren't using a separate `lxc` inet table, and then setting up the non-namespaced chains inside that table (with netfilter hooks), rather than using the `filter` table?

This would potentially make removing the rules even easier as you can just delete the whole table.

The nice thing is that you can use the `add` command and if the entity already exists it won't error, so that allows you to always `add` the table and chains you need, and if they don't exist they'll get setup.

I've used this effect when I wanted to ensure that the chains exist in order to flush their rules in a single call to nftables:

https://github.com/lxc/lxd/blob/master/lxd/firewall/drivers/drivers_nftables_templates.go#L101-L109

pablofsf replied:
Thank you very much for the comments. You're totally right, I assumed that the `nat` and `filter` tables existed, my bad. And had not thought about creating a separate table, which definitely makes total sense. I'll fix that up.

tomponline commented on config/init/common/lxc-net.in:
```diff
@@ -51,6 +55,52 @@ _ifup() {
+        $NFT add rule ip6 nat lxc-nat counter masquerade
+        $NFT add rule ip6 nat postrouting ip saddr ${LXC_IPV6_NETWORK} ip daddr != ${LXC_IPV6_NETWORK} jump lxc-nat
+    fi
+    $NFT add chain inet filter lxc-filter
```
It might be worth trying to utilise nftable's ability to atomically add multiple rules at once, to avoid multiple calls to nftables and to avoid leaving the firewall in an inconsistent state due to partial failure.

For an example of this see https://github.com/lxc/lxd/blob/master/lxd/firewall/drivers/drivers_nftables.go#L589

pablofsf replied:
Should be fixed now!

tomponline reviewed (changes requested):
I've added a few of things to think about, but I think the main issue at the moment is that this script appears to assume the `filter` and `nat` tables exists  (unless I've missed something) which cannot be guaranteed and would prevent the script from running.

pablofsf commented on config/init/common/lxc-net.in:
```diff
@@ -51,6 +55,52 @@ _ifup() {
+    $NFT add rule inet filter lxc-filter accept
+    $NFT add rule inet filter input iifname ${LXC_BRIDGE} udp dport { 53, 67 } jump lxc-filter
+    $NFT add rule inet filter input iifname ${LXC_BRIDGE} tcp dport { 53, 67 } jump lxc-filter
+    $NFT add rule inet filter forward iifname ${LXC_BRIDGE} jump lxc-filter
```
I'm using `ip` and `ip6` families for the `nat` table, whose hook is of type `nat`. If I understand the documentation correctly, `nat` is only supported by `ip` and `ip6`  families, but not by `inet`. 
Is there something I'm misunderstanding? https://wiki.nftables.org/wiki-nftables/index.php/Quick_reference-nftables_in_10_minutes#Chains

tomponline replied:
The documentation does suggest that, but I'm actively using this on focal 20.04's nftables:

```
table inet tomp {
        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;

                # Outbound NAT.
                oifname $ext_if ip saddr $int_net4 snat $ext_v4
                oifname $modem_if ip saddr $int_net4 ip daddr 192.168.2.1 masquerade
        }
}
```

tomponline replied:
Or to put it another way I've only got `table inet tomp` and inside that I've got both snat, masquerade and dnat rules working fine.

tomponline replied:
So it might make it easier to just have one inet table called `lxc` and put everything you need inside it.

pablofsf replied:
Well, that might actually be some Ubuntu specific kernel patch, or maybe only available after a certain version. Running your script with the variables defined:
```
~$ sudo nft -c -f test-nft
test-nft:11:52-55: Error: NAT is only supported for IPv4/IPv6
                oifname $ext_if ip saddr $int_net4 snat $ext_v4
                                                   ^^^^
```

pablofsf replied:
I'm running Debian buster:
```
~$ uname -a
Linux kgpe-d16 4.19.0-16-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
```

tomponline replied:
I think its that the kernel version of nftables is older, in LXD we only target kernel >= 5.2 due to several missing features and incompatibilities in nftables until that point. It depends how older version of nftables you're targetting.

tomponline replied:
See https://github.com/lxc/lxd/pull/8725

tomponline replied:
Before that we were targetting >= 5.0 for a different reason, but I forget what that was.

pablofsf replied:
Then I guess there's a bug in the nftables online documentation, which should definitely inform about that. If you think that would work, then I have no problem implementing it that way, but it should be informed to users and we should add an extra check. Otherwise adding two tables for the nat rules shouldn't be much of a hazard

tomponline replied:
Ah, there was this comment:

https://github.com/lxc/lxd/commit/5118984d3d4b1a03e0867314d3eeb5f159387fdb#diff-390446f8c9cd49885f7ad1d2f9fe64044d1a404a4501286dbb98ba5d9d4eec58L51

"// We require a 5.x kernel to avoid weird conflicts with xtables."

tomponline replied:
Yeah maybe check the kernel is >= 5.2, or go with separate tables for nat, which ever you feels is better for your needs. If you want to align with LXD then >= 5.2 check is the way to go.

pablofsf replied:
For me, I'd be much happier using different tables. Since LXC rules are definitely simpler than LXD rules, hopefully that does not become an issue

tomponline replied:
Cool, in that case you may be better off just using `ip` and `ip6` `lxc` tables and then duplicating the small amount of inbound rules you need for each protocol.

pablofsf commented:
@tomponline So second iteration. Definitely looks better now!

tomponline commented on config/init/common/lxc-net.in:
```diff
@@ -51,6 +55,58 @@ _ifup() {
+    NFT_RULESET=""
+    if [ -n "$LXC_IPV6_ARG" ] && [ "$LXC_IPV6_NAT" = "true" ]; then
+        NFT_RULESET="${NFT_RULESET}
+add table ip6 lxc-nat6;
```
@pablofsf you could just use the same table name `lxc` for all of the 3 table types (`ip`, `ip6` and `inet`), it looks like nft allows this and would mean the firewall only uses 1 namespace.

pablofsf replied:
Fixed!

tomponline commented:
This is looking a lot cleaner  now thanks :)

pablofsf commented:
Thank you for your time and your feedback!

tomponline commented on config/init/common/lxc-net.in:
```diff
@@ -26,8 +27,11 @@ LXC_IPV6_NAT="false"
-use_iptables_lock="-w"
-iptables -w -L -n > /dev/null 2>&1 || use_iptables_lock=""
+NFT="$(which nft)"
+if [ -z "$NFT" ] || ! nft list ruleset > /dev/null 2>&1 || [ "$LXC_USE_NFT" != "true" ]; then
```
@pablofsf it feels like this line and the similar two lines that are `[ -n "$NFT" ] && nft list ruleset > /dev/null 2>&1 && [ "$LXC_USE_NFT" = "true" ];` could do with being moved into a function called something like `useNftables()` that way the logic governing the use of nftables is kept in one place and it would read a bit easier when used.

tomponline reviewed (commented):
Thanks, just a minor nit now and I think we are then ready :)

pablofsf commented:
Yeah, I should have taken care of that one from the beginning. Now the three checks just call one function

tomponline reviewed (approved):
Looks good thanks.

@stgraber please could you make a quick pass over this to check the general style as I've not used lxc-net before so would appreciate your input.

stgraber commented:
jenkins: test this please
